sklearn/datasets/_openml.py

```diff
@@ -448,27 +448,83 @@ def _get_num_samples(data_qualities):
-    # production!
-    url = _DATA_FILE.format(file_id)
 
-    @_retry_with_clean_cache(url, data_home)
-    def _arff_load():
-        with closing(_open_openml_url(url, data_home)) as response:
-            if sparse is True:
-                return_type = _arff.COO
+    if sparse is True:
+        return_type = _arff.COO
+    else:
+        return_type = _arff.DENSE_GEN
+
+    response = _open_openml_url(url, data_home)
+
+    with closing(response):
+        # Note that if the data is dense, _arff.load will parse the headers,
+        # but the data will only be read line by line when it is converted
+        # below. This is intended to minimise memory usage.
+        arff = _arff.load((line.decode('utf-8')
+                           for line in response),
+                          return_type=return_type,
+                          encode_nominal=not as_frame)
+
+        frame = nominal_attributes = None
+        if as_frame:
+            columns = data_columns + target_columns
+            frame = _convert_arff_data_dataframe(arff, columns, features_dict)
+            X = frame[data_columns]
+            if len(target_columns) >= 2:
+                y = frame[target_columns]
+            elif len(target_columns) == 1:
+                y = frame[target_columns[0]]
             else:
-                return_type = _arff.DENSE_GEN
-
-            arff_file = _arff.loads(response.read().decode('utf-8'),
-                                    encode_nominal=encode_nominal,
-                                    return_type=return_type)
-        return arff_file
-
-    return _arff_load()
+                y = None
+        else:
+            X, y = _convert_arff_data(arff['data'], col_slice_x,
+                                      col_slice_y, shape)
+
+            # nominal attributes is a dict mapping from the attribute name to
+            # the possible values. Includes also the target column (which will
+            # be popped off below, before it will be packed in the Bunch
+            # object)
+            nominal_attributes = {k: v for k, v in arff['attributes']
+                                  if isinstance(v, list) and
+                                  k in data_columns + target_columns}
+            is_classification = {col_name in nominal_attributes
+                                 for col_name in target_columns}
+            if not is_classification:
+                # No target
+                pass
+            elif all(is_classification):
+                y = np.hstack([
+                    np.take(
+                        np.asarray(nominal_attributes.pop(col_name),
+                                   dtype='O'),
+                        y[:, i:i + 1].astype(int, copy=False))
+                    for i, col_name in enumerate(target_columns)
+                ])
+            elif any(is_classification):
+                raise ValueError('Mix of nominal and non-nominal targets is '
```

Comment:
This `ValueError` can still be passed through to the retry handler.

Maybe we let ValueError's not trigger a retry?

Reply:
Perhaps you're right. I just recalled that ARFF parsing issues should result in a non-ValueError.